src/be_db_block.erl

```diff
@@ -12,138 +12,200 @@
+            [
+                "insert into transactions (block, time, hash, type, fields) values ($1, $2, $3, $4, $5) ",
+                "on conflict do nothing"
+            ],
+            []
+        ),
 
     #{
-      ?S_BLOCK_HEIGHT => S0,
-      ?S_INSERT_BLOCK => S1,
-      ?S_INSERT_BLOCK_SIG => S2,
-      ?S_INSERT_TXN => S3
-     }.
+        ?S_BLOCK_HEIGHT => S0,
+        ?S_INSERT_BLOCK => S1,
+        ?S_INSERT_BLOCK_SIG => S2,
+        ?S_INSERT_TXN => S3
+    }.
 
 %%
 %% be_block_handler
 %%
 
 init(_) ->
     {ok, _, [{Value}]} = ?PREPARED_QUERY(?S_BLOCK_HEIGHT, []),
-    Height = case Value of
-                 null -> 0;
-                 _ -> Value
-             end,
+    Height =
+        case Value of
+            null -> 0;
+            _ -> Value
+        end,
     lager:info("Block database at height: ~p", [Height]),
     {ok, #state{
-            height = Height
-           }}.
+        height = Height
+    }}.
 
-load_block(Conn, Hash, Block, _Sync, Ledger, State=#state{}) ->
+load_block(Conn, Hash, Block, _Sync, Ledger, State = #state{}) ->
     BlockHeight = blockchain_block_v1:height(Block),
-    ?assertEqual(BlockHeight, State#state.height + 1,
-                 "New block must line up with stored height"),
+    ?assertEqual(
+        BlockHeight,
+        State#state.height + 1,
+        "New block must line up with stored height"
+    ),
     %% Seperate the queries to avoid the batches getting too big
     BlockQueries = q_insert_block(Hash, Block, Ledger, [], State),
     ok = ?BATCH_QUERY(Conn, BlockQueries),
-    {ok, State#state{height=BlockHeight}}.
+    maybe_write_snapshot(Block, blockchain_worker:blockchain()),
+    {ok, State#state{height = BlockHeight}}.
 
 %%
 %% API
 %%
 
-block_height(#state{height=Height}) ->
+block_height(#state{height = Height}) ->
     Height.
 
-
 %%
 %% Internal
 %%
 
-q_insert_block(Hash, Block, Ledger, Queries, State=#state{base_secs=BaseSecs}) ->
+maybe_write_snapshot(Height, Chain) when is_number(Height) ->
+    {ok, Block} = blockchain:get_block(Height, Chain),
+    maybe_write_snapshot(Block, Chain);
+maybe_write_snapshot(Block, Chain) ->
+    Height = blockchain_block_v1:height(Block),
+    try
+        maybe_write_snapshot(
+            Height,
+            blockchain_block_v1:snapshot_hash(Block),
+            os:getenv("SNAPSHOT_DIR"),
+            Chain
+        )
+    catch
+        What:Why:Where ->
+            lager:warning("Failed to write snapshot ~p: ~p", [Height, {What, Why, Where}])
+    end.
+
+maybe_write_snapshot(_, _, false, _Chain) ->
+    ok;
+maybe_write_snapshot(_, <<>>, _, _Chain) ->
+    ok;
+maybe_write_snapshot(Height, SnapshotHash, SnapshotDir, Chain) ->
+    {ok, Snapshot} = blockchain:get_snapshot(SnapshotHash, Chain),
+    Filename = filename:join([SnapshotDir, io_lib:format("snap-~p", [Height])]),
+    {ok, BinSnap} = blockchain_ledger_snapshot_v1:serialize(Snapshot),
+    ok = file:write_file(Filename, BinSnap).
```

Comment:
will this end up being a directory that fills up forever? note for our future selves that we should have disk monitoring.

Reply:
heh... yes it will